tests/nn/data_parallel/test_fsdp_multiple_forward.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-class-docstring
+# pylint: disable=missing-function-docstring
+
+""" Test FSDP with different multiple forward of the same module. """
+
+import tempfile
+
+import pytest
+import torch
+import torch.multiprocessing as mp
+from torch.nn import Linear, Module
+from torch.optim import SGD
+
+from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
+from fairscale.nn.data_parallel import TrainingState
+from fairscale.utils.testing import dist_init, skip_if_single_gpu, teardown, torch_version
+
+
+def _test_func(rank, world_size, fsdp_config, tempfile_name, unused):
+    result = dist_init(rank, world_size, tempfile_name, unused)
+    assert result, "Dist init failed"
+
+    assert isinstance(fsdp_config, dict), str(fsdp_config)
+
+    class Model(Module):
+        def __init__(self):
+            super().__init__()
+            self.inner = FSDP(Linear(4, 4), **fsdp_config)
+            self.outer = Linear(4, 5)
+
+        def forward(self, x):
+            # Forward twice.
+            i = self.inner(x)
+            j = self.inner(x)
+            return self.outer(i + j)
+
+    model = FSDP(Model(), **fsdp_config).cuda()
+    optim = SGD(model.parameters(), lr=0.1)
+
+    for _ in range(3):
+        in_data = torch.rand(64, 4).cuda()
+        in_data.requires_grad = True
+        out = model(in_data)
+        out.sum().backward()
+        optim.step()
+        optim.zero_grad()
+
+    model.assert_state(TrainingState.IDLE)
```

Comment:
is it worth testing that the results are identical to DDP too? I do like the cleanliness of this test though 😄

Reply:
Yeah, I thought about it. My plan is to have another test with double-forward and conv/bn etc with numerical testing with DDP. This test is only for hooks I guess. :-)